### DISCUSSION TITLE
Is the computation of Snell's law correct?

### DISCUSSION THREAD

**Discussion by ligoudaner377** [mitsuba-renderer/mitsuba3#825] [Q&A] [answered]
https://github.com/mitsuba-renderer/mitsuba3/blob/a946691a0d5272a80ea45f7b5f22f31d697cf290/include/mitsuba/render/fresnel.h#L160

I have difficulty understanding the above code for computing the cos value of the transmitted light.
To me,
`Float cos_theta_t_sqr =
        dr::fnmadd(dr::fnmadd(cos_theta_i, cos_theta_i, -1.f), eta_ti * eta_ti, 1.f);`
is more meaningful.

Sorry in advance. if I asked a too-naive question

**Comment by njroussel** [comment, marked as answer]
Hi @ligoudaner377 

By definition we have `sin_theta_t = et_ti * sin_theta_i`
The current code translates to:
```
-(-sqr(cos_theta_i) + 1) * sqr(eta_ti) + 1 =  # which we'll rewrite as
1 - (1 - sqr(cos_theta_i)) * sqr(eta_ti) =    # using trigonometric identities we get
1 - sqr(sin_theta_i) * sqr(eta_ti) =          # if we take the square of our definition and subsitute it in this formula we get
1 - sqr(sin_theta_t) =                        # again with the trigonometric identity:
sqr(cos_theta_t)
```

Do you agree?

**Reply by ligoudaner377** [reply]
I see, I confused the functions fnmadd and fmadd.
Thank you!!
